### DISCUSSION TITLE
How to resume training with correct learning rate

### DISCUSSION THREAD

**Discussion by RamakrishnaChaitanya** [idiap/coqui-ai-TTS#326] [Q&A] [answered]
I'm training a fastpitch model from scratch on custom datasets. The initial learning rate (lr) that was set prior to training is 0.0001 i.e., 

`    "lr": 0.0001,
    "optimizer": "Adam",
    "optimizer_params": {
        "betas": [
            0.9,
            0.998
        ],
        "weight_decay": 1e-06
    },
    "lr_scheduler": "NoamLR",
    "lr_scheduler_params": {
        "warmup_steps": 4000
    },`

But the learning rate that got printed at the start of training was quite different i.e., 2.5000000000000002e-08 . 

`[4m[1m > EPOCH: 0/1500[0m
 --> /data/Ramakrishna/Projects/TTS/Github_repos/coqui-ai-TTS/recipes/vctk/fast_pitch/fast_pitch_indic-March-03-2025_04+49PM-a2fb3662

[1m > TRAINING (2025-03-03 16:52:55) [0m

[1m   --> TIME: 2025-03-03 16:53:00 -- STEP: 0/686 -- GLOBAL_STEP: 0[0m
     | > loss_spec: 29.786182403564453  (29.786182403564453)
     | > loss_dur: 0.1483408361673355  (0.1483408361673355)
     | > loss_pitch: 0.051399536430835724  (0.051399536430835724)
     | > loss_aligner: 10.167981147766113  (10.167981147766113)
     | > loss_binary_alignment: 0.29392918944358826  (0.29392918944358826)
     | > loss: 40.44783401489258  (40.44783401489258)
     | > duration_error: 4.2506937980651855  (4.2506937980651855)
     | > grad_norm: tensor(72.2665, device='cuda:0')  (tensor(72.2665, device='cuda:0'))
     | > **current_lr: 2.5000000000000002e-08** 
     | > step_time: 3.0655  (3.0655248165130615)
     | > loader_time: 1.4772  (1.4772047996520996)`

I assume that it is happening due to the usage of warmup steps, but I'm not very sure. And later, due to Out of Memory exception, the training got stopped after 100 epochs. I would like to resume training the model with the updated learning rate (lr) that was used in 100th epoch i.e., 6.1 * e-6. However, when i tried executing this command 

`CUDA_VISIBLE_DEVICES="7" python train_fastpitch.py \
    --restore_path  path/model_file.pth
    --coqpit.lr 6.1*e-4`

The learning rate that got printed in the latest epoch is again quite different i.e., 2.5000000000000002e-08. Can i get any help regarding how to resume the training with the updated learning rate?

**Comment by eginhard** [comment, marked as answer]
To continue a training run, you should use `--continue_path path/` instead of `--restore_path path/model.pth`, that should correctly restore parameters.

**Reply by RamakrishnaChaitanya** [reply]
Thanks!

**Reply by RamakrishnaChaitanya** [reply]
@eginhard If we use the --continue_path, the initial learning rate is getting used instead of the updated learning rate. For example. the initial learning rate that i was using is **0.0001** and i'm using **Noam scheduler**. Hence, in the first epoch, I'm observed the **current_lr** as **2.5 * e^-8**. And, i trained the fastpitch model for 2000 epochs. And at the end of 2000 epochs, the learning rate that got printed in the logs was **5.73 * e^-5.** 

Later, i wanted to continue training the model for another 2000 epochs. When i executed the training command with --continue path, the learning rate that got printed in the first epoch of fresh run was again **2.5 * e^-8 instead of 5.73 * e^-5**. I tried changing the **lr** parameter and **warmup_steps** to 1 in the config.json manually, but that didn't work out. I also tried updating the **lr** and **warmup_steps** in the **default** fastpitch_config.json, but even that didn't work out.

**Reply by eginhard** [reply]
Thank you for investigating and clarifying! I just released coqui-tts-trainer version 0.3.0, which includes many fixes w.r.t. restoring models. Most importantly, the LR scheduler `state_dict` is now also added to the checkpoints, so that when using `--continue_path` you'll continue exactly where you stopped (this will only work for new checkpoints saved in the updated version).

You can already test this with the `dev` branch, or otherwise once the next version 0.27.0 of coqui-tts gets released (1-2 weeks probably).

**Comment by RamakrishnaChaitanya** [comment]
@eginhard If we use the --continue_path, the initial learning rate is getting used instead of the updated learning rate. For example. the initial learning rate that i was using is 0.0001 and i'm using Noam scheduler. Hence, in the first epoch, I'm observed the current_lr as 2.5 * e^-8. And, i trained the fastpitch model for 2000 epochs. And at the end of 2000 epochs, the learning rate that got printed in the logs was 5.73 * e^-5.

Later, i wanted to continue training the model for another 2000 epochs. When i executed the training command with --continue path, the learning rate that got printed in the first epoch of fresh run was again 2.5 * e^-8 instead of 5.73 * e^-5. I tried changing the lr parameter and warmup_steps to 1 in the config.json manually, but that didn't work out. I also tried updating the lr and warmup_steps in the default fastpitch_config.json, but even that didn't work out.

Obsered that initial step is 0 in the fresh run, which ideally shouldn't happen. 

![image](https://github.com/user-attachments/assets/26b48b04-cb7f-4742-9266-b5cf2f80dbac)
